src/wlapplication_mousewheel_options.cc

```diff
@@ -0,0 +1,323 @@
+		if (can_handle(x, y, modstate)) {
+			return Vector2i(x * current_sign_x_, y * current_sign_y_);
+		}
+		return Vector2i(0, 0);
+	}
+};
+
+static const std::map<MousewheelOptionID, MousewheelOption> mousewheel_options = {
+   {MousewheelOptionID::kUIChangeValueInvertX,
+    MousewheelOption("change_value_x_invert", MousewheelOptionType::kBool, false, false)},
+   {MousewheelOptionID::kUIChangeValueInvertY,
+    MousewheelOption("change_value_y_invert", MousewheelOptionType::kBool, false, false)},
+
+   {MousewheelOptionID::kUITabInvertX,
+    MousewheelOption("tabpanel_x_invert", MousewheelOptionType::kBool, false, false)},
+   {MousewheelOptionID::kUITabInvertY,
+    MousewheelOption("tabpanel_y_invert", MousewheelOptionType::kBool, false, false)},
+
+   {MousewheelOptionID::kMapZoomMod,
+    MousewheelOption("zoom_modifier", MousewheelOptionType::kKeymod, KMOD_NONE, KMOD_CTRL)},
+   {MousewheelOptionID::kMapZoomX,
+    MousewheelOption("zoom_x", MousewheelOptionType::kBool, true, true)},
+   {MousewheelOptionID::kMapZoomY,
+    MousewheelOption("zoom_y", MousewheelOptionType::kBool, true, true)},
+   {MousewheelOptionID::kMapZoomInvertX,
+    MousewheelOption("zoom_x_invert", MousewheelOptionType::kBool, false, false)},
+   {MousewheelOptionID::kMapZoomInvertY,
+    MousewheelOption("zoom_y_invert", MousewheelOptionType::kBool, false, false)},
+
+   {MousewheelOptionID::kMapScrollMod,
+    MousewheelOption("move_map_modifier", MousewheelOptionType::kKeymod, KMOD_NONE, KMOD_NONE)},
+   {MousewheelOptionID::kMapScroll,
+    MousewheelOption("move_map", MousewheelOptionType::kBool, false, true)},
+
+   {MousewheelOptionID::kGameSpeedMod,
+    MousewheelOption("gamespeed_modifier", MousewheelOptionType::kKeymod, KMOD_ALT, KMOD_ALT)},
+   {MousewheelOptionID::kGameSpeedX,
+    MousewheelOption("gamespeed_x", MousewheelOptionType::kBool, true, true)},
+   {MousewheelOptionID::kGameSpeedY,
+    MousewheelOption("gamespeed_y", MousewheelOptionType::kBool, true, true)},
+
+   {MousewheelOptionID::kEditorToolsizeMod,
+    MousewheelOption(
+       "editor_toolsize_modifier", MousewheelOptionType::kKeymod, KMOD_ALT, KMOD_ALT)},
+   {MousewheelOptionID::kEditorToolsizeX,
+    MousewheelOption("editor_toolsize_x", MousewheelOptionType::kBool, true, true)},
+   {MousewheelOptionID::kEditorToolsizeY,
+    MousewheelOption("editor_toolsize_y", MousewheelOptionType::kBool, true, true)},
+
+   {MousewheelOptionID::kAlwaysOn, MousewheelOption("", MousewheelOptionType::kBool, true, true)},
+   {MousewheelOptionID::kDisabled, MousewheelOption("", MousewheelOptionType::kBool, false, false)},
+   {MousewheelOptionID::kNoMod,
+    MousewheelOption("", MousewheelOptionType::kKeymod, KMOD_NONE, KMOD_NONE)},
+
+   {MousewheelOptionID::kUse2Ddefaults,
+    MousewheelOption("use_2d_defaults", MousewheelOptionType::kBool, false, true)},
+
+};
+
+// Default signs
+#define SIGN_INCREASE_RIGHT -1
```

Comment:
Please don't use `ifdef`s for constants, use a constexpr instead:
```c++
constexpr int kSignIncreaseRight = -1;
```

Reply:
Yes, but these are only used in the `define`-s of x-y pairs for the initialisation of `mousewheel_handlers`, and never used alone. Should they outlive the pair `define`-s? Or should the pairs be turned into constexpr vectors too, and the constructor of `MousewheelHandlerOptions` be made to take them apart into `default_sign_x_` and `default_sign_y_`? Do we want these constexpr-s in the global namespace?

Reply:
`#define`s are bad style and should be avoided where possible – just replace them with some `constexpr`s in the .cc file. +1 for using `vector`, or even better, `pair`:
```c++
using Sign2D = std::pair<int, int>;
/* Replace int variables in MousewheelHandlerOptions with Sign2D */

constexpr int kSignIncreaseRight = -1;
...
constexpr Sign2D kDefaultSignValue(kSignIncreaseRight, kSignIncreaseUp);
...
```

Reply:
Hmmm, I think I'll use Vector2i, because it has members x and y instead of first and second...

Reply:
OK, it seems on macos pair is defined with const&... I guess I could use `struct Sign2D { int x, y; };`?

Reply:
>  or not... clang-tidy says it's not a constant expression...

+1 for Vector2i – you can use `static const` instead of `constexpr` if the compiler has hiccups.